cli/__init__.py

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+# PYTHON_ARGCOMPLETE_OK
+
+"""
+An installer for Foreman.
+"""
+import os
+import obsah
+import sys
+
+
+class ApplicationConfig(obsah.ApplicationConfig):
+    """
+    A class describing the where to find various files
+    """
+
+    @staticmethod
+    def name():
+        """
+        Return the name as shown to the user in the ArgumentParser
+        """
+        return 'rop'
+
+    @staticmethod
+    def data_path():
+        """
+        Return the data path. Houses playbooks and configs.
+        """
+        path = os.environ.get('ROP_DATA')
+        if path is None:
+            path = os.getcwd()
+
+        return path
+
+    @classmethod
+    def ansible_config_path(cls):
+        """
+        Return the ansible.cfg path
+        """
+        return os.environ.get('OBSAH_ANSIBLE_CFG', os.path.join(cls.data_path(), 'ansible.cfg'))
+
+    @staticmethod
+    def inventory_path():
+        """
+        Return the inventory path
+        """
+        return os.environ.get('OBSAH_INVENTORY', os.path.join(os.getcwd(), 'inventories'))
+
+
+def main(cliargs=['--help'], application_config=ApplicationConfig):  # pylint: disable=R0914
+    """
+    Main command
+    """
+    if len(sys.argv) > 1:
+        cliargs = sys.argv[1:]
+
```

Comment:
It's a bad practice in Python to have lists as defaults because they're shared between invocations. In this particular case it's not too bad but if we end up adding linters they may flag it so it's best to just avoid it:
```suggestion
def main(cliargs=None, application_config=ApplicationConfig):  # pylint: disable=R0914
    """
    Main command
    """
    if not cliargs:
        cliargs = sys.argv[1:]
        if not cliargs:
            cliargs = ['--help']
```

Perhaps this is a bit safer?

Though I really wonder if this shouldn't be the default obsah behavior. From my understanding of argparse and how obsah calls it I'd think this should work:
```suggestion
def main(cliargs=None, application_config=ApplicationConfig):  # pylint: disable=R0914
    """
    Main command
    """
```

Reply:
Should work in theory, that's how Obal works. But for my environment and testing, it does not work unless I have what I have currently.

Reply:
Can you explain a bit better what you mean by "does not work"?

I have applied the following patch here locally:
```
diff --git cli/__init__.py cli/__init__.py
index 7b57828..9af357c 100755
--- cli/__init__.py
+++ cli/__init__.py
@@ -47,12 +47,10 @@ class ApplicationConfig(obsah.ApplicationConfig):
         return os.environ.get('OBSAH_INVENTORY', os.path.join(os.getcwd(), 'inventories'))
 
 
-def main(cliargs=['--help'], application_config=ApplicationConfig):  # pylint: disable=R0914
+def main(cliargs=None, application_config=ApplicationConfig):  # pylint: disable=R0914
     """
     Main command
     """
-    if len(sys.argv) > 1:
-        cliargs = sys.argv[1:]
 
     obsah.main(cliargs=cliargs, application_config=application_config)
```
and `rop` seems to work fine?

```
% ./rop              
usage: rop [-h] action ...
rop: error: the following arguments are required: action

% ./rop -h
usage: rop [-h] action ...

positional arguments:
  action         which action to execute
    deploy       Install
    pull-images  Pull necessary container images
    setup        Setup up a host to be installed.

options:
  -h, --help     show this help message and exit

% ./rop deploy -h
usage: rop deploy [-h] [-v] [-e EXTRA_VARS] [--candlepin-database-name CANDLEPIN_DATABASE_NAME] [--candlepin-database-password CANDLEPIN_DATABASE_PASSWORD] [--candlepin-database-user CANDLEPIN_DATABASE_USER]
                  [--certificate-source {default,installer}] [--database-host DATABASE_HOST] [--database-mode {internal,external}] [--database-port DATABASE_PORT] [--database-ssl-ca DATABASE_SSL_CA]
                  [--database-ssl-mode DATABASE_SSL_MODE] [--foreman-database-name FOREMAN_DATABASE_NAME] [--foreman-database-password FOREMAN_DATABASE_PASSWORD] [--foreman-database-user FOREMAN_DATABASE_USER]
                  [--pulp-database-name PULP_DATABASE_NAME] [--pulp-database-password PULP_DATABASE_PASSWORD] [--pulp-database-user PULP_DATABASE_USER]

Install

options:
  -h, --help            show this help message and exit
  -v, --verbose         verbose output
  --candlepin-database-name CANDLEPIN_DATABASE_NAME
                        Name of the Candlepin database.
…
```

Reply:
But obal doesn't do that either

(I agree it would be cool, but it works as the other thing right now)

Reply:
Yea, obal doesn't work that way either but that doesn't mean it shouldn't :) And right now this version of it does work that way... what's the downside of this implementation?

Reply:
Well, you wrote "Should work in theory, that's how Obal works." which led me to assume "it works in obal but not in rop".

The downside is mostly "it's ugly" :)

As an alternative (and that would benefit all consumers), we could adjust it in obsah, like this:
```
diff --git obsah/__init__.py obsah/__init__.py
index 08b919d..52546da 100755
--- obsah/__init__.py
+++ obsah/__init__.py
@@ -316,6 +316,8 @@ def obsah_argument_parser(application_config=ApplicationConfig, playbooks=None,
     if argcomplete:
         argcomplete.autocomplete(parser)
 
+    parser.format_usage = parser.format_help
+
     return parser
```
(this forces "usage" which is what's used here, to use the longer "help" output instead)

I poked a few Pythonistas if there is a nicer way.

Reply:
What @evgeni suggested is also where I was heading, just didn't have time to look at the code just yet.

Reply:
That sounds like a winner!